File: Practica4/QytetetRuby P4/pkg/QytetetRuby-0.0.1/lib/calle.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'titulo_propiedad'
module ModeloQytetet

class Calle < Casilla
  
  attr_accessor :num_hoteles, :num_casas, :titulo
  attr_reader :coste
  
  def initialize(num_ca,cos,tip,titulo)
    super(num_ca,tip)
    @coste = cos
    @titulo = titulo
    @num_hoteles = 0
    @num_casas = 0
    titulo.casilla = self
  end
  
  
  def asignar_propietario(jugador)
      @titulo.propietario = jugador
      @titulo
  end
  
   def calcular_valor_hipoteca
      hipoteca_base = @titulo.hipoteca_base
      cantidad_recibida = hipoteca_base + (@num_casas*0.5*hipoteca_base + @num_hoteles*hipoteca_base).round
      cantidad_recibida
    end
    
    def cancelar_hipoteca
      coste_alquiler_base = @titulo.alquiler_base
      cantidad_recibida = coste_alquiler_base + (@num_casas * 0.5 + @num_hoteles * 2).round
      cantidad_recibida
    end
    
    def cobrar_alquiler
      coste_alquiler_base = @titulo.alquiler_base
      coste_alquiler = coste_alquiler_base + (@num_casas*0.5 + @num_hoteles*2).round
      @titulo.cobrar_alquiler(coste_alquiler)
      
      coste_alquiler
    end
    
    def edificar_casa
      @num_casas = @num_casas + 1
      coste_edificar_casa = @titulo.precio_edificar
      coste_edificar_casa
    end
    
    def edificar_hotel
      @num_casas = 0
      @num_hoteles = @num_hoteles + 1
      coste_edificar_hotel = @titulo.precio_edificar
      coste_edificar_hotel

    end
    
    def esta_hipotecada
      @titulo.hipotecada
    end
    
    def hipotecar
      @titulo.hipotecada = true
      calcular_valor_hipoteca()
    end
    
    def precio_total_comprar
      return nil
    end
    
    def propietario_encarcelado
      @titulo.propietario_encarcelado()
    end
    
    def se_puede_edificar_casa(factor_especulador)
      if @num_casas < (4 * factor_especulador)
        return true
      else
        return false
      end
    end
    
    def se_puede_edificar_hotel (factor_especulador)
      if @num_casas == (4 * factor_especulador) && @num_hoteles < (4*factor_especulador)
        return true
      else
        return false
      end
      
    end
    
    
    
    def tengo_propiedad
      @titulo.tengo_propietario
    end
    
    def get_precio
      @titulo.precio_edificar
    end
    
    def vender_titulo
      precio_compra = @coste + (@num_casas + @num_hoteles * @titulo.precio_edificar)
      precio_venta = (precio_compra + @titulo.factor_revalorizacion * precio_compra).round 
    
      @titulo.propietario = nil
      @num_casas = 0
      @num_hoteles = 0
      
      precio_venta
    end
    
    def asignar_titulo_propiedad
      return nil
    end
  
  
  
  # Preguntar
#    private :titulo= 
    private :asignar_titulo_propiedad
  
end
end

File: Practica5/QytetetJava/src/modeloqytetet/Calle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author Lorena
 */

public class Calle extends Casilla{
    //Variables
    private int coste; //indica el coste de compra del titulo de la propiedad de esa casilla
    private int numHoteles, numCasas; // indica el numero de hoteles y de casas edificados en esa casilla
    private TituloPropiedad titulo; //para asociar la casilla a su titulo de propiedad
    
    //Constructor
    Calle(){}
    Calle(int numeroCasilla, int coste, TipoCasilla tipo, TituloPropiedad titulo){
        super(numeroCasilla, tipo);
        this.coste = coste;
        numHoteles = 0;
        numCasas = 0;
        this.titulo = titulo;
        
       titulo.setCasilla(this);
    }
    
    //Modificadores
    void setNumHoteles(int nuevoNumero){
        numHoteles = nuevoNumero;
    }
    
    void setNumCasas(int nuevoNumero){
        numCasas = nuevoNumero;
    }
    
    private void setTitulo(TituloPropiedad titulo){
        this.titulo = titulo;
    }
    
    //Consultores
    int getCoste(){
        return coste;
    }
    
    int getNumHoteles(){
        return numHoteles;
    }
    
    int getNumCasas(){
        return numCasas;
    }
    
    public TituloPropiedad getTitulo(){
        return titulo;
    }
    
    int getCosteHipoteca(){
        return titulo.getHipotecaBase();
    }
    
    int getPrecioEdificar(){
        return titulo.getPrecioEdificar();
    }
    
    //Métodos
    
    TituloPropiedad asignarPropietario(Jugador jugador){
        titulo.setPropietario(jugador);
        return titulo;
    }
    
    
    int calcularValorHipoteca(){
        int hipotecaBase = titulo.getHipotecaBase();
        int cantidadRecibida = hipotecaBase + (int)(this.getNumCasas()*0.5*hipotecaBase + this.getNumHoteles()*hipotecaBase);
        return cantidadRecibida;
    }
    
    
    int cancelarHipoteca(){
        int cantidadRecibida = (int)(this.calcularValorHipoteca() + 0.1);
        return cantidadRecibida;
    }
    
    
    int cobrarAlquiler(){
        int costeAlquilerBase = titulo.getAlquilerBase();
        int costeAlquiler = costeAlquilerBase + (int)(this.getNumCasas()*0.5 + this.getNumHoteles()*2);
        
        titulo.cobrarAlquiler(costeAlquiler);
        
        return costeAlquiler;
    }
    
    
    int edificarCasa(){
        int nuevoNumero = this.getNumCasas() + 1;
        this.setNumCasas(nuevoNumero);
        
        int costeEdificarCasa = titulo.getPrecioEdificar();
        return costeEdificarCasa;
    }
    
    
    int edificarHotel(){
        setNumCasas(0);
        int nuevoNumero = this.getNumHoteles() + 1;
        this.setNumHoteles(nuevoNumero);
        
        int costeEdificarHotel = titulo.getPrecioEdificar();
        return costeEdificarHotel;
    }
    
    
    boolean estaHipotecada(){
        return titulo.getHipotecada();
    }
    
    
    int hipotecar(){
        titulo.setHipotecada(true);
        return this.calcularValorHipoteca();
    }
    
    
    int precioTotalComprar(){
        throw new UnsupportedOperationException("Sin implementar");
    }
    
    
    boolean propietarioEncarcelado(){
        return titulo.propietarioEncarcelado();
    }
    
    
    public boolean sePuedeEdificarCasa(int factorEspeculador){
        return getNumCasas() < 4*factorEspeculador;
    }
    
    
    public boolean sePuedeEdificarHotel(int factorEspeculador){
        if ( getNumCasas() == 4*factorEspeculador && getNumHoteles() < 4*factorEspeculador){
            return true;
        }
        else{
            return false;
        }
    }
       
    boolean tengoPropietario(){
        return titulo.tengoPropietario();
    }
    
    
    int venderTitulo(){
        int precioCompra = this.getCoste() + (this.getNumCasas() + this.getNumHoteles()) * titulo.getPrecioEdificar();
        int precioVenta = (int)(precioCompra + titulo.getFactorRevalorizacion()*precioCompra);
        
        titulo.setPropietario(null);
        this.setNumHoteles(0);
        this.setNumCasas(0);
        
        return precioVenta;
    }
    
    
    protected void asignarTituloPropiedad(){}
    
        //Método toString() devuelve un String con el estado del objeto correspondiente
    
    public String toString(){
        return "\n Numero de la Casilla = " + Integer.toString(super.getNumeroCasilla())
                + "\n Coste = " + Integer.toString(coste)
                + "\n Numero de hoteles = " + Integer.toString(numHoteles) 
                + "\n Numero de casas = " + Integer.toString(numCasas)
                + "\n Tipo de la casilla = " + super.getTipo()
                + "\n Titulo de propiedad = " + titulo.toString() + "\n";
    }

    
}
